### PR TITLE
Use EmailFilter in Email Sender

### PR DESCRIPTION
**Pull request by kanwalgupta**
# Description

This PR includes the following proposed change(s):

-TCVP 840

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring / Documentation
- [ ] Version change

if your change is a breaking change, please add `breaking change` label to this PR

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Does the change impact or break the Docker build?

- [ ] Yes
- [x] No

If Yes: Has Docker been updated accordingly?

- [ ] Yes
- [ ] No

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/backend/TrafficCourtsApi/ticket-worker/Program.cs [hunk 1]
```diff
@@ -42,7 +44,13 @@ internal static void ConfigureServiceBus(IServiceCollection services, IConfigura
             var port = configuration.GetSection("SMTPServer")["Port"];
             var from = configuration.GetSection("Mail")["From"];
             var fromEmail = configuration.GetSection("Mail")["FromEmail"];
+            string allowedEmails = configuration.GetSection("ALLOWED_RECIPIENT_EMAILS").Value;
 
+            if(allowedEmails != "*")
+            {
+                var allowedEmailsList = allowedEmails?.Split(";").ToList();
+                services.AddSingleton<IEmailFilter>(new DefaultEmailFilter(allowedEmailsList?? new List<string>()));
+            }
```
Comment (peggy-gov-bc-ca) [inline src/backend/TrafficCourtsApi/ticket-worker/Program.cs:53]: If it is *, need to add NotFilteredEmailFilter to service collection.

File: src/backend/TrafficCourtsApi/ticket-worker/Features/Emails/EmailSender.cs [hunk 2]
```diff
@@ -84,17 +92,20 @@ public async Task SendUsingTemplate(string to, string subject, TicketDisputeCont
...
+                var email = _email
                     .To(to)
                     .Subject(subject)
-                    .UsingTemplateFromEmbedded("ticket-worker.Features.Emails.Resources.submissiontemplate.liquid", emailModel, this.GetType().GetTypeInfo().Assembly, true)
-                    .SendAsync();
+                    .UsingTemplateFromEmbedded("ticket-worker.Features.Emails.Resources.submissiontemplate.liquid", emailModel, this.GetType().GetTypeInfo().Assembly, true);
 
-                
-                if (!result.Successful)
+                if ( _emailFilter == null || _emailFilter.IsAllowed(to))
```
Comment (peggy-gov-bc-ca) [inline src/backend/TrafficCourtsApi/ticket-worker/Features/Emails/EmailSender.cs:100]: _emailFilter == null  cannot be null as in contractor, it is already protected it cannot be null.

Comment (peggy-gov-bc-ca) [reply src/backend/TrafficCourtsApi/ticket-worker/Features/Emails/EmailSender.cs:100]: Can we move _emailFilter.IsAllowed to line 91?

**Review by peggy-gov-bc-ca — CHANGES_REQUESTED** [review]

File: src/backend/TrafficCourtsApi/ticket-worker/Features/Emails/EmailSender.cs (hunk 2 above)
Comment (pbolduc) [reply src/backend/TrafficCourtsApi/ticket-worker/Features/Emails/EmailSender.cs:100]: > Can we move _emailFilter.IsAllowed to line 91?

My intention is to still format the email, even though sending is not allowed.

File: src/backend/TrafficCourtsApi/ticket-worker/Features/Emails/EmailSender.cs [hunk 3]
```diff
@@ -84,17 +92,20 @@ public async Task SendUsingTemplate(string to, string subject, TicketDisputeCont
...
                 {
-                    _logger.LogError("Failed to send an email");
+                    var result = await email.SendAsync();
+                    if (!result.Successful)
+                    {
+                        _logger.LogError("Failed to send an email");
+                    }
                 }
+
```
Comment (pbolduc) [inline src/backend/TrafficCourtsApi/ticket-worker/Features/Emails/EmailSender.cs:108]: I would also like log messages to indicate that the target email address was not allowed to be sent.  In addition, we should have a debug log message to indicate that the IsAllowed block returned true.

File: src/backend/TrafficCourtsApi/ticket-worker/Program.cs [hunk 4]
```diff
@@ -42,7 +44,13 @@ internal static void ConfigureServiceBus(IServiceCollection services, IConfigura
             var port = configuration.GetSection("SMTPServer")["Port"];
             var from = configuration.GetSection("Mail")["From"];
             var fromEmail = configuration.GetSection("Mail")["FromEmail"];
+            string allowedEmails = configuration.GetSection("ALLOWED_RECIPIENT_EMAILS").Value;
 
+            if(allowedEmails != "*")
```
Comment (pbolduc) [inline src/backend/TrafficCourtsApi/ticket-worker/Program.cs:49]: Why is configuring the email settings in ConfigureServiceBus?

File: src/backend/TrafficCourtsApi/ticket-worker/Features/Emails/EmailSender.cs [hunk 5]
```diff
@@ -84,17 +92,20 @@ public async Task SendUsingTemplate(string to, string subject, TicketDisputeCont
...
 
-                
-                if (!result.Successful)
+                if ( _emailFilter == null || _emailFilter.IsAllowed(to))
                 {
-                    _logger.LogError("Failed to send an email");
+                    var result = await email.SendAsync();
+                    if (!result.Successful)
+                    {
+                        _logger.LogError("Failed to send an email");
```
Comment (pbolduc) [inline src/backend/TrafficCourtsApi/ticket-worker/Features/Emails/EmailSender.cs:105]: Be sure to log out the error messages. It wont be very helpful if all we know it failed but dont have any additional details.

File: src/backend/TrafficCourtsApi/ticket-worker/Program.cs [hunk 6]
```diff
@@ -42,7 +44,13 @@ internal static void ConfigureServiceBus(IServiceCollection services, IConfigura
             var port = configuration.GetSection("SMTPServer")["Port"];
             var from = configuration.GetSection("Mail")["From"];
             var fromEmail = configuration.GetSection("Mail")["FromEmail"];
+            string allowedEmails = configuration.GetSection("ALLOWED_RECIPIENT_EMAILS").Value;
```
Comment (pbolduc) [inline src/backend/TrafficCourtsApi/ticket-worker/Program.cs:47]: On all these configuration items, we need to be more defensive, what if something is not configured? Or configured wrong? Should the app just crash? Should it log out that there is an error with the email configuration and emails will be disabled? Should it disable the handler?

**Review by pbolduc — CHANGES_REQUESTED** [review]

**Review by peggy-gov-bc-ca — APPROVED** [review]

**Pull request merged**